tools/web-test-runner-helpers.js

```diff
@@ -1,12 +1,172 @@
+		if (elem.waitForIt !== undefined) {
+			await elem.waitForIt();
+		}
+		await update;
+		await nextFrame();
+		hasUpdateComplete = true;
+	}
+
+	const children = getComposedChildren(elem);
+	//console.log(`${spaces}${elem.tagName} ${hasUpdateComplete} ${children?.length || 0}`);
+	if (children !== null) {
+		await Promise.all(children.map(e => waitForElem(e, myDepth)));
+	}
+}
+
+export const fixture = async(element, opts) => {
+	await Promise.all([reset(opts), waitForFonts()]);
+	const elem = await wcFixture(element);
+	await waitForElem(elem, -1);
+	return elem;
+};
 
 export const focusWithKeyboard = async(element) => {
-	await sendKeys({ press: 'Tab' });
+	await sendKeys({ press: 'Escape' });
 	element.focus({ focusVisible: true });
 };
 
 export const focusWithMouse = async(element) => {
-	const { x, y } = element.getBoundingClientRect();
-	await sendMouse({ type: 'click', position: [Math.ceil(x), Math.ceil(y)] });
+	const { height, width, x, y } = element.getBoundingClientRect();
+	await sendMouse({ type: 'click', position: [Math.ceil(x + width / 2), Math.ceil(y + height / 2)] });
 	await sendMouse({ type: 'move', position: [0, 0] });
 };
+
+export const hoverWithMouse = async(element) => {
+	mouseResetNeeded = true;
+	const { height, width, x, y } = element.getBoundingClientRect();
+	await sendMouse({ type: 'move', position: [Math.ceil(x + width / 2), Math.ceil(y + height / 2)] });
+};
+
+async function reset(opts) {
+
+	opts = opts || {};
+	opts.lang = opts.lang || DEFAULT_LANG;
+	opts.rtl = opts.lang.startsWith('ar') || !!opts.rtl;
+	opts.viewport = opts.viewport || {};
+	opts.viewport.height = opts.viewport.height || DEFAULT_VIEWPORT_HEIGHT;
+	opts.viewport.width = opts.viewport.width || DEFAULT_VIEWPORT_WIDTH;
+
+	let awaitNextFrame = false;
+
+	window.scroll(0, 0);
+
+	if (mouseResetNeeded) {
+		mouseResetNeeded = false;
+		await sendMouse({ type: 'move', position: [0, 0] });
+	}
+
+	if (opts?.theme !== currentTheme) {
+		if (opts?.theme !== undefined) {
+			document.body.setAttribute('data-theme', opts.theme);
+		} else {
+			document.body.removeAttribute('data-theme');
+		}
+		awaitNextFrame = true;
+		currentTheme = opts?.theme;
+	}
+
+	if (opts.rtl !== currentRtl) {
+		document.documentElement.setAttribute('dir', opts.rtl ? 'rtl' : 'ltr');
+		awaitNextFrame = true;
+		currentRtl = opts.rtl;
+	}
+
+	if (opts.lang !== currentLang) {
+		getDocumentLocaleSettings().language = opts.lang;
+		currentLang = opts.lang;
+		awaitNextFrame = true;
+	}
+
+	if (opts.viewport.height !== currentViewportHeight || opts.viewport.width !== currentViewportWidth) {
+		await setViewport(opts.viewport);
+		awaitNextFrame = true;
+		currentViewportHeight = opts.viewport.height;
+		currentViewportWidth = opts.viewport.width;
+	}
+
+	// TODO: reset focus
+
+	if (awaitNextFrame) {
+		await nextFrame();
+	}
+
+}
+
+export const screenshotAndCompare = async(elem, name, opts) => {
+
+	const rect = elem.getBoundingClientRect();
```

Comment:
We currently do some fancy logic to handle getting the proper `rect` for dropdowns (and filter, etc.): https://github.com/BrightspaceUI/core/blob/main/components/dropdown/test/dropdown-helper.js#L28-L44
Do we see that being added here, or providing users a way to override? Or should they create a "large enough" wrapper element?

Reply:
Making a wrapper element that is "large enough" is the simplest, although adjusting the hard-coded dimensions any time the contents change is a bit of an annoyance. It can provide some degree of buffer when contents change that would normally leave us without a diff due to the size changing, though my understanding is that we're going to improve that aspect. I tend to like being able to override with something more automatic like this.

Reply:
Having said that, the consumer could always have some script that resizes the wrapper before the screenshot.

Reply:
What if they could do something like this in `fixture`:

```javascript
const elem = await fixture(html`<d2l-my-elem>...</d2l-my-elem>`, {
  margin: {
    blockEnd: 200,
    inlineStart: 50,
    inlineEnd: 75
  }
});
```

Then we'd reserve `200px` at the bottom, `50px` to the left and `75px` to the right in the form of padding on the body (we default to `30px` today). Alternately we could just put a huge amount of padding on the body -- like `500px` -- but then the viewport size would need to change.

So that would reserve the space initially as it renders. Then when they take the screenshot maybe it can reuse that value when it captures the rectangle? `screenshotAndCompare` already supports a single-value "margin" option (defaults to `10px` all around).

Reply:
I think that would work if the consumer hard-codes it.

Seems like that wouldn't work if the consumer wants to avoid hard-coding - if they wanted to measure the opener + the content to figure out the size needed (like the helper Stacey linked), it needs to have access to the element, but it won't exist yet. 🐔 & 🥚 ... 😄 . It would need a way to adjust margin after it's been rendered. Or, for the rare cases where a consumer might want this, they just include a wrapper div in their fixture and they set the dimensions of it based on measurement just before doing the screenshot.